xarray/coding/times.py

```diff
@@ -669,8 +663,10 @@ def encode_cf_datetime(
         time_deltas = dates_as_index - ref_date
 
         # Use floor division if time_delta evenly divides all differences
-        # to preserve integer dtype if possible (GH 4045).
-        if np.all(time_deltas % time_delta == np.timedelta64(0, "ns")):
+        # to preserve integer dtype if possible (GH 4045)
+        # NaT prevents us from using datetime64 directly, but we can safely coerce
+        # to int64 in presence of NaT, so we just dropna before check.
+        if np.all(time_deltas.dropna() % time_delta == np.timedelta64(0, "ns")):
```

Comment:
Note to self, .dropna() removes any NaT from time_deltas. As consequence if-clause returns True if int64 can be preserved. Unfortunately the floor division still returns float64 (as we have NaT inside). We would need to do something like:

```python
num = time_deltas.astype(np.int64) // time_delta.astype(np.int64)
```
to keep int64. IMHO we can do this as `NaT` translates to `-9223372036854775808` (minimal int64) and back. It might be possible to special case NaT to use it as FillValue.

Reply:
This check has been removed. It's tackled by ref-dat and unit comparison.